### PR TITLE
Create a bash script to provision your backstage's namespace

### PR DESCRIPTION
**Pull request by cmoulliard**
- Created a bash script to provision your backstage's namespace 
- See: #173

### CODE REVIEW

**Review by aureamunoz — APPROVED** [review]
LGTM

File: README.md [hunk 1]
```diff
@@ -115,9 +115,26 @@ This section explains how to use Backstage:
...
+```bash
+./scripts/provision-namespace.sh -h
+Processing -h
+Usage: ./scripts/provision-namespace.sh [options]
+Options:
+  -n, --namespace     <namespace>                        The namespace on the QShift cluster (mandatory)
+  -q, --quay          <quay_username:quay_password>      The Quay username and password (mandatory)
+  -o, --quay-org      <quay_username:quay_password>      The Quay organization hosting the images (mandatory)
+  -d, --docker        <docker_username:docker_password>  The docker username and password (mandatory)
+  -k, --key-path      <public_key_path>                  The path of your public to ssh to the VM (mandatory)
```
Comment (aureamunoz) [inline README.md]:
```suggestion
  -k, --key-path      <public_key_path>                  The path to your ssh public key for the VM (mandatory)
```

File: README.md [hunk 2]
```diff
@@ -115,9 +115,26 @@ This section explains how to use Backstage:
...
+  -r, --dry-run                                          Run the kubectl command with dry-run=client
+```
```
Comment (aureamunoz) [inline README.md]:
```suggestion
For example:

./scripts/provision-namespace.sh -n my-namespace -q "my-quay-user:my-quay-pwd" -o "my-quay-org" -d "my-docker-user:my-docker-pwd" -k $HOME/.ssh/id_rsa.pub
```

**Review by aureamunoz — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (cmoulliard) [reply README.md]: There is mistake with your example. That should be 
```
./scripts/provision-namespace.sh -n my-namespace -q "my-quay-user:my-quay-pwd" -o "my-quay-org" -d "my-docker-user:my-docke-pwd" -k $HOME/.ssh/id_rsa.pub
```

Comment (aureamunoz) [reply README.md]: ok, I have replaced my suggestion, I think you can commit it.

Comment (cmoulliard) [reply README.md]: Still a typo mistake "my-docke-rpwd" -> my-docker-pwd"

Comment (aureamunoz) [reply README.md]: done

Comment (cmoulliard) [reply README.md]: Your commit suggestion is still 
```
./scripts/provision-namespace.sh -n my-namespace -q "my-quay-user:my-quay-pwd" -o "my-quay-org" -d "my-docker-user:my-docke-rpwd" -k $HOME/.ssh/id_rsa.pub
```
and not 
```
./scripts/provision-namespace.sh -n my-namespace -q "my-quay-user:my-quay-pwd" -o "my-quay-org" -d "my-docker-user:my-docker-pwd" -k $HOME/.ssh/id_rsa.pub
```

Comment (aureamunoz) [reply README.md]: I think it's ok now, if not you can apply the correct one from your side.

**Review by aureamunoz — APPROVED** [review]

File: scripts/provision-namespace.sh [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+# Display usage
+function usage() {
+  echo "Usage: $0 [options]"
+  echo "Options:"
+  echo "  -n, --namespace     <namespace>                        The namespace on the QShift cluster (mandatory)"
+  echo "  -q, --quay          <quay_username:quay_password>      The Quay username and password (mandatory)"
+  echo "  -o, --quay-org      <quay_username:quay_password>      The Quay organization hosting the images (mandatory)"
+  echo "  -d, --docker        <docker_username:docker_password>  The docker username and password (mandatory)"
+  echo "  -k, --key-path      <public_key_path>                  The path of your public to ssh to the VM (mandatory)"
```
Comment (iocanel) [inline scripts/provision-namespace.sh]: This should be optional and default to $HOME/.ssh/id_rsa.pub

File: scripts/provision-namespace.sh [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+# Set default values
+DRY_RUN=false
+
+# Display usage
+function usage() {
+  echo "Usage: $0 [options]"
+  echo "Options:"
+  echo "  -n, --namespace     <namespace>                        The namespace on the QShift cluster (mandatory)"
+  echo "  -q, --quay          <quay_username:quay_password>      The Quay username and password (mandatory)"
+  echo "  -o, --quay-org      <quay_username:quay_password>      The Quay organization hosting the images (mandatory)"
```
Comment (iocanel) [inline scripts/provision-namespace.sh]: This is confusing. What do I need to add here? org, username ? password ?

File: scripts/provision-namespace.sh [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  echo "  -d, --docker        <docker_username:docker_password>  The docker username and password (mandatory)"
```
Comment (iocanel) [inline scripts/provision-namespace.sh]: I shouldn't be needing both quay and docker. One of the two should do.

**Review by iocanel — CHANGES_REQUESTED** [review]
Added some minor comments related to the parameters and their defaults.

File: scripts/provision-namespace.sh (hunk 5 above)
Comment (iocanel) [reply scripts/provision-namespace.sh]: Also, both docker and quay can be read from `$HOME/.docker/config.json` we should try that if no parameter has been specified.

File: scripts/provision-namespace.sh (hunk 3 above)
Comment (cmoulliard) [reply scripts/provision-namespace.sh]: That could not be optional for 2 reasons:
- As images are pulled, we can then reach the famous docker limit and by consequence docker creds are needed
- As we will certainly demo/test within the provisioned namespace an application generated which then triggers argocd => tekton pipeline where an image is build, then such an image will be pushed right now on your quay account. 

I agree to default `$HOME/.ssh/id_rsa.pub`

File: scripts/provision-namespace.sh (hunk 4 above)
Comment (cmoulliard) [reply scripts/provision-namespace.sh]: The quay.io registry allows to push images under different organizations and creds can be defined at the root of quay.io or quay.io/ORG => 
```
{
  "auths": {
    "quay.io/$QUAY_ORG": {
      "auth": "$QUAY_CREDS_BASE64"
    },
```
By doing that you have the guarantee that the token/creds matches the ORG level where such a robot token has been defined

`Robot Accounts are named tokens that can be granted permissions on multiple repositories under this organization. They are typically used in environments where credentials will be shared, such as deployment systems. `

File: scripts/provision-namespace.sh (hunk 5 above)
Comment (cmoulliard) [reply scripts/provision-namespace.sh]: Good idea to also pass as parameter the config.json file.
 
Remark: Such a file to be imported will not work for macos users as docker manage the creds using OSKEYSTORE
```
cat ~/.docker/config.json
{
	"auths": {
		"ghcr.io": {},
		"https://index.docker.io/v1/": {},
		"quay.io": {}
	},
	"credsStore": "osxkeychain",
```

File: scripts/provision-namespace.sh (hunk 4 above)
Comment (cmoulliard) [reply scripts/provision-namespace.sh]: I reviewed the wording. See commit : f6298acf3276fe82b64c14af88eddb90f53e7f96

**Comment by cmoulliard** [comment]
I improved the script. Here is what you will see when you provide the mandatory arguments
```
./bin/provision-namespace.sh -n cmoullia -q "qshift+github:xxxxxxx" -o ch007m -d "ch007m@gmail.com:yyyyyy" --dry-run
Let's creating the resources
Using as values ...
Your namespace=cmoullia
Quay registry username=qshift+github
Quay registry password=********
Quay registry organization=ch007m
Dockerhub registry username=ch007m@gmail.com
Dockerhub registry password=********
Public key path=/Users/cmoullia/.ssh/id_rsa.pub
.....
### Creating new project for: cmoullia
oc new-project cmoullia 
### Creating container registry secret ...
kubectl --dry-run=client create secret generic dockerconfig-secret --from-file=config.json 
secret/dockerconfig-secret created (dry run)

### Creating the backstage service account
kubectl --dry-run=client create sa my-backstage 
serviceaccount/my-backstage created (dry run)

Give cluster-admin role to the backstage SA to access the kubernetes API resources
kubectl --dry-run=client apply -f rbac.yml 
clusterrolebinding.rbac.authorization.k8s.io/backstage-cmoullia-cluster-access configured (dry run)

### Creating a secret hosting your SSH public key
kubectl --dry-run=client create secret generic quarkus-dev-ssh-key --from-file=key=/Users/cmoullia/.ssh/id_rsa.pub 
secret/quarkus-dev-ssh-key created (dry run)

### Creating the Fedora podman virtual machine
kubectl --dry-run=client apply -f manifest/installation/virt/quarkus-dev-virtualmachine.yml 
virtualmachine.kubevirt.io/quarkus-dev configured (dry run)

### Apply the label: argocd.argoproj.io/managed-by=<argocd_namespace> on your namespace
kubectl --dry-run=client label namespace cmoullia argocd.argoproj.io/managed-by=openshift-gitops 
namespace/cmoullia not labeled (dry run)

### Adding your namespace to the parameter: sourceNamespaces of the ArgoCD CR ...
Namespace 'cmoullia' already exists in sourceNamespaces.
```

**Comment by cmoulliard** [comment]
I squashed the commits into one and rebased on main. WDYT ? @iocanel @aureamunoz

**Comment by cmoulliard** [comment]
FYI: I did a test on a mac to create the registry creds from the docker config file but that don't work as mentioned before
```
kubectl create secret docker-registry dummy-creds --from-file=.dockerconfigjson=$HOME/.docker/config.json
secret/dummy-creds created

BUT the content imported cannot be read AND creds are not there as managed by osxkeychain

.dockerconfigjson: "{\n\t\"auths\": {\n\t\t\"ghcr.io\": {},\n\t\t\"gitea.cnoe.localtest.me:8443\":
  {\n\t\t\t\"auth\": \"Z2l0ZWFBZG1pbjo/RSckOEU8Ril2PG0zZi8zX1YmJGxYMmEvNTRdPzRxaj9xam4xYURr\"\n\t\t},\n\t\t\"https://index.docker.io/v1/\":
  {},\n\t\t\"quay.io\": {}\n\t},\n\t\"credsStore\": \"osxkeychain\",\n\t\"currentContext\":
  \"desktop-linux\",\n\t\"plugins\": {\n\t\t\"debug\": {\n\t\t\t\"hooks\": \"exec\"\n\t\t},\n\t\t\"scout\":
  {\n\t\t\t\"hooks\": \"pull,buildx build\"\n\t\t}\n\t},\n\t\"features\": {\n\t\t\"hooks\":
  \"true\"\n\t}\n}"
"
```

If now, we decide to use and support only podman, then we can grab such credentials from the file AND use such a file (if it exists as default with the command to avoid to have to provide the quay.io, dockerhub, etc credentials)
```
kubectl create secret docker-registry dummy-creds --from-file=.dockerconfigjson=~/.config/containers/auth.json
...
.dockerconfigjson: |-
  {
          "auths": {
                  "docker.io": {
                          "auth": "Y2*********************Uc3"
                  },
                  "gitea.cnoe.localtest.me:8443": {
                          "auth": "Z2l0Z*******************ZF00OSZx"
                  },
                  "quay.io": {
                          "auth": "Y2**************************R1A="
                  }
          }
  }
```
WDYT ? @iocanel @aureamunoz

**Review by iocanel — APPROVED** [review]
LGTM

**Pull request merged**